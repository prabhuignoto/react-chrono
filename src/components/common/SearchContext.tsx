import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
