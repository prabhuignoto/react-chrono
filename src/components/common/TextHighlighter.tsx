import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
