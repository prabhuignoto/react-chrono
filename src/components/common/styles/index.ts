import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
