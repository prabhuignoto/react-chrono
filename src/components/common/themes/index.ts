import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
