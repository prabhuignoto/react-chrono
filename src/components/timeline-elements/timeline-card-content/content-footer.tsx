import React, {
  FunctionComponent,
  RefObject,
  useContext,
  useMemo,
} from 'react';
import { Theme } from '../../../models/Theme';
import { TimelineMode } from '../../../models/TimelineModel';
import { GlobalContext } from '../../GlobalContext';
import ChevronIcon from '../../icons/chev-right';
import {
  ChevronIconWrapper,
  ShowMore,
  SlideShowProgressBar,
  TriangleIconWrapper,
} from './timeline-card-content.styles';

export type ContentFooterProps = {
  canShow: boolean;
  onExpand: () => void;
  paused: boolean;
  progressRef: RefObject<HTMLDivElement>;
  remainInterval: number;
  showMore: boolean;
  showProgressBar?: boolean;
  showReadMore?: boolean | '';
  startWidth: number;
  textContentIsLarge: boolean;
  theme?: Theme;
  triangleDir?: string;
};

const ContentFooter: FunctionComponent<ContentFooterProps> = ({
  theme,
  showProgressBar,
  onExpand,
  triangleDir,
  showMore,
  textContentIsLarge,
  showReadMore,
  remainInterval,
  paused,
  startWidth,
  canShow,
  progressRef,
}) => {
  const { mode } = useContext(GlobalContext);

  const canShowTriangleIcon = useMemo(() => {
    return (['VERTICAL', 'VERTICAL_ALTERNATING'] as TimelineMode[]).some(
      (m) => m === mode,
    );
  }, [mode]);

  return (
    <>
      {showReadMore && textContentIsLarge ? (
        <ShowMore
          className="show-more"
          onClick={onExpand}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              onExpand();
            }
          }}
          role="button"
          show={canShow}
          theme={theme}
          tabIndex={0}
        >
          {<span>{showMore ? 'read less' : 'read more'}</span>}
          <ChevronIconWrapper collapsed={!showMore}>
            <ChevronIcon />
          </ChevronIconWrapper>
        </ShowMore>
      ) : null}

      {showProgressBar && (
        <SlideShowProgressBar
          color={theme?.primary}
          duration={remainInterval}
          paused={paused}
          ref={progressRef}
          startWidth={startWidth}
        ></SlideShowProgressBar>
      )}

      {canShowTriangleIcon && (
        <TriangleIconWrapper
          dir={triangleDir}
          theme={theme}
        ></TriangleIconWrapper>
      )}
    </>
  );
};

export { ContentFooter };
