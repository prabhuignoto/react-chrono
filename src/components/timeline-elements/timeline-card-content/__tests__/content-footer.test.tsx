import userEvent from '@testing-library/user-event';
import { createRef } from 'react';
import { describe, it, vi } from 'vitest';
import { customRender } from '../../../common/test';
import { providerProps } from '../../../common/test/index';
import { ContentFooter } from '../content-footer';
import { ContentFooterProps } from '../header-footer.model';

const commonProps: ContentFooterProps = {
  canShow: false,
  onExpand: () => {},
  paused: false,
  progressRef: createRef(),
  remainInterval: 0,
  showMore: false,
  showProgressBar: false,
  showReadMore: false,
  startWidth: 0,
  textContentIsLarge: false,
  triangleDir: 'up',
};

describe('ContentFooter', () => {
  //should match snapshot
  it('should match snapshot', () => {
    const { container } = customRender(
      <ContentFooter
        {...commonProps}
        startWidth={300}
        showProgressBar
        paused={false}
      />,
      {
        providerProps,
      },
    );

    expect(container).toMatchSnapshot();
  });

  // should check if the progress bar is rendered with a with of 100
  it('should check if the progress bar is rendered with a width of 100', () => {
    const { getByRole } = customRender(
      <ContentFooter
        {...commonProps}
        showProgressBar
        startWidth={100}
        paused
      />,
      {
        providerProps,
      },
    );

    expect(getByRole('progressbar')).toBeInTheDocument();
    expect(getByRole('progressbar')).toBeVisible();

    // expect(getByRole('progressbar')).toHaveStyle('width: 100px');
  });

  //should render read more button when text content is large
  it('should render read more button when text content is large', () => {
    const { getByText } = customRender(
      <ContentFooter
        {...commonProps}
        showReadMore
        textContentIsLarge
        canShow
      />,
      {
        providerProps,
      },
    );

    expect(getByText('read more')).toBeInTheDocument();
    expect(getByText('read more')).toBeVisible();
  });

  //should render read less button when text content is large
  it('should render read less button when text content is large', () => {
    const { getByText } = customRender(
      <ContentFooter
        {...commonProps}
        showReadMore
        textContentIsLarge
        canShow
        showMore
      />,
      {
        providerProps,
      },
    );

    expect(getByText('read less')).toBeInTheDocument();
    expect(getByText('read less')).toBeVisible();
  });

  //should call onExpand when read more button is clicked
  it('should call onExpand when read more button is clicked', async () => {
    const onExpand = vi.fn();
    const user = userEvent.setup();

    const { getByText } = customRender(
      <ContentFooter
        {...commonProps}
        showReadMore
        textContentIsLarge
        canShow
        onExpand={onExpand}
      />,
      {
        providerProps,
      },
    );

    await user.click(getByText('read more'));
    expect(onExpand).toHaveBeenCalled();
  });
});
