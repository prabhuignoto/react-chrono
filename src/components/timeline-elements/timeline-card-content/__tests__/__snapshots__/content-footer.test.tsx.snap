// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContentFooter > should match snapshot 1`] = `
<div>
  <progress
    class="sc-fmKFGs cIUnCn"
    color="#0f52ba"
  />
  <span
    class="sc-blmEgr kkdMbx"
    dir="up"
    offset="-8"
  />
</div>
`;
